src/routes/devices.ts

```diff
@@ -569,67 +634,77 @@ export const state: RequestHandler = (req, res) => {
 								}
 							})
 							.then(() => {
-								_.each(device.manages__device as AnyObject[], depDev => {
+								_.each(device.manages__device, depDev => {
 									dependent.devices[depDev.uuid] = {
 										name: depDev.device_name,
 										apps: {
-											[depDev.belongs_to__application[0].id]: {
+											[getExpanded(depDev.belongs_to__application)!.id]: {
 												config: {},
 												environment: {},
 											},
 										},
 									};
-									const app: AnyObject =
+									const app =
 										dependent.devices[depDev.uuid].apps[
-											depDev.belongs_to__application[0].id
+											getExpanded(depDev.belongs_to__application)!.id
 										];
-									varListInsert(depDev.device_config_variable, app.config);
+									varListInsert(depDev.device_config_variable!, app.config);
 
-									release = depReleases[depDev.belongs_to__application[0].id];
+									release =
+										depReleases[
+											getExpanded(depDev.belongs_to__application)!.id
+										];
 
-									const image = _.get(release, 'contains__image[0].image[0]');
+									const image = _.get(
+										release,
+										'contains__image[0].image[0]',
+									) as Image;
 									const svcInstall =
 										image == null
 											? null
 											: serviceInstallFromImage(depDev, image);
 
 									if (image != null) {
 										varListInsert(
-											image.image_environment_variable,
+											// TODO: investigate whether this is a bug
+											((image as any) as NonNullable<
```

Comment:
@CameronDiver can you clarify whether this is correct?
I think that image_environment_variable doesn't have an FK to an image
```
Fact type: image is part of release
	Synonymous Form: release contains image
	Term Form: release image

	Fact type: release image has name (Auth)
		Term Form: image environment variable
		Database Table Name: image environment variable
```
![image](https://user-images.githubusercontent.com/1295829/57526957-7384f680-7337-11e9-8bf8-c47f98872dcf.png)

Reply:
Hmm I'm not sure the image environment variables are actually used anywhere, but the FK is from `image__is_part_of__release` and `image_environment_variable`

Reply:
Fixed in https://github.com/balena-io/open-balena-api/pull/109

Reply:
@Page- how about the rest of the PR ? :)
Should I rebase or do you prefer to close it?

Reply:
I think it's worthwhile to rebase (although it might be easier to redo the types on the state endpoint specifically because it's gone through a lot of changes), I would ditch cases where we know the expand is safe though, eg for
```ts
depDev.belongs_to__application[0].id
```
vs
```ts
getExpanded(depDev.belongs_to__application)!.id
```
I'd keep the former rather than introducing the `!`

Reply:
@Page- maybe we could do something like
```
type PropsOfType<T, P> = {
	[K in keyof T]: T[K] extends P ? K : never;
}[keyof T];

type AssociatedResource<T = WithId> =
	| NavigationResource<T>
	| ReverseNavigationResource<T>;

type ExpandableProps<T> = PropsOfType<T, AssociatedResource>;

export type ExpandedProps<T, U extends ExpandableProps<T>> = T & Required<Pick<T, U>>;
```
?

Reply:
I mean, since the `Device` props are typed with `NavigationResource` and `ReverseNavigationResource` we somehow have to tell TS that `depDev` has them expanded...
On the other hand, if we get those types in PineClient, then we could have pine itself return the proper result by inspecting the provided props.
The SDK atm does check such props but just for validating the PineOptions. The next step of course is to return a result that is exactly typed as the PineOptions.